### PR TITLE
Add Homogenizer model as comparison model

### PR DESCRIPTION
**Pull request by amywinecoff**
We have already established several baselines (random, ideal) that are useful for evaluation in the Chaney replication. The Homogenizer class is another model class that can serve as both a comparison point and additional mechanism to ensure internal validity (i.e., that the simulation is behaving as expected). It will validate that homogenization can occur within the simulation since the Homogenizer model explicitly homogenizes the algorithms' user representation over time. 

The class takes on the behavior of another model (e.g., content, MF) during startup and train to establish individual `users_hat`. Then the model creates a single representation (i.e., fully homogenized representation) for all users. The value for `users_hat` is gradually updated so that it is a weighted average of the individual users' representation and the homogenized representation where more weight is given to the homogenized representation over time until its weight reaches 1. You can specify how quickly this happens through the parameters `homogenization_increment`

The inheritance gymnastics executed here may not be the best way to accomplish the goal of the class, so please let me know if there is a more desirable way to set this up.

### CODE REVIEW

File: trecs/models/homogenizer.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+        homogenization_increment=0.1,
+        model_class=ContentFiltering,
+        **kwargs
+    ):
+
+        self.__class__ = type(
+            self.__class__.__name__, (model_class, object), dict(self.__class__.__dict__)
+        )
+
+        # measurements = [MSEMeasurement()]
```
Comment (sunnymatt) [inline trecs/models/homogenizer.py:147]: Let's just delete this line!

File: trecs/models/homogenizer.py [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+            actual_user_representation,
+            actual_item_representation,
+            probabilistic_recommendations,
+            seed,
+            verbose,
+            num_items_per_iter,
+            **kwargs
+        )
+
+        assert 0 < homogenization_increment < 1
```
Comment (sunnymatt) [inline trecs/models/homogenizer.py:164]: can we turn this into an `if` statement that explicitly raises an exception if the condition is not satisfied?

File: trecs/models/homogenizer.py [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+        gen = Generator(seed=seed)
+
+        self.homogenized_users_hat = np.tile(
+            gen.normal(size=(1, num_attributes)), (self.num_users, 1)
+        )
```
Comment (sunnymatt) [inline trecs/models/homogenizer.py:170]: this is super neat!

File: trecs/models/homogenizer.py [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+                An array of item indices that users have interacted with in the
+                latest step. Namely, `interactions[u]` represents the index of
+                the item that the user has interacted with.
+
+        """
+        interactions_per_user = np.zeros((self.num_users, self.num_items))
+        interactions_per_user[self.users.user_vector, interactions] = 1
+        user_attributes = np.dot(interactions_per_user, self.items_hat.T)
+
+        self.individual_users_hat += user_attributes
```
Comment (sunnymatt) [inline trecs/models/homogenizer.py:199]: correct me if I'm wrong, but this updates the individual user representation the same way that our ContentFiltering model does, right?

**Review by sunnymatt — CHANGES_REQUESTED** [review]
This is great - I'm really happy we have this in the codebase now! 2 minor points of feedback:
- Since my new changes for sparse matrices, this will probably need to be edited to reflect the new changes
- We should add tests for exactly the above reason^ that is, when other parts of the codebase change, tests will ensure that we don't accidentally break the Homogenizer functionality!

@amywinecoff let me know if you'd like me to take lead on updating the PR with the new changes from the last PR and to write tests. Apologies for the huge delay in reviewing!

**Comment by amywinecoff** [comment]
Based on our Slack conversation, this model should go in the t-recs-experiment repo, so I am going to close this PR. I will address these comments within the t-recs-experiments repo when I use the model.

**Pull request closed without merging**